tensorboard/data/server/server.rs

```diff
@@ -247,26 +247,208 @@ impl TensorBoardDataProvider for DataProviderHandler {
+                        _ => continue,
+                    };
+                run_res
+                    .tags
+                    .push(data::list_blob_sequences_response::TagEntry {
+                        tag_name: tag.0.clone(),
+                        metadata: Some(data::BlobSequenceMetadata {
+                            max_step: max_step.into(),
+                            max_wall_time: max_wall_time.into(),
+                            max_length: max_length as i64,
+                            summary_metadata: Some(*ts.metadata.clone()),
+                        }),
+                    });
+            }
+            if !run_res.tags.is_empty() {
+                run_res.run_name = run.0.clone();
+                res.runs.push(run_res);
+            }
+        }
+
+        Ok(Response::new(res))
     }
 
     async fn read_blob_sequences(
         &self,
-        _request: Request<data::ReadBlobSequencesRequest>,
+        req: Request<data::ReadBlobSequencesRequest>,
     ) -> Result<Response<data::ReadBlobSequencesResponse>, Status> {
-        Err(Status::unimplemented("not yet implemented"))
+        let req = req.into_inner();
+        let want_plugin = parse_plugin_filter(req.plugin_filter)?;
+        let (run_filter, tag_filter) = parse_rtf(req.run_tag_filter);
+        let num_points = parse_downsample(req.downsample)?;
+        let runs = self.read_runs()?;
+
+        let mut res: data::ReadBlobSequencesResponse = Default::default();
+        for (run, data) in runs.iter() {
+            if !run_filter.want(run) {
+                continue;
+            }
+            let data = data
+                .read()
+                .map_err(|_| Status::internal(format!("failed to read run data for {:?}", run)))?;
+            let mut run_res: data::read_blob_sequences_response::RunEntry = Default::default();
+            for (tag, ts) in &data.blob_sequences {
+                if !tag_filter.want(tag) {
+                    continue;
+                }
+                if plugin_name(&ts.metadata) != Some(&want_plugin) {
+                    continue;
+                }
+
+                let mut points = ts.valid_values().collect::<Vec<_>>();
+                downsample::downsample(&mut points, num_points);
+                let n = points.len();
+                let mut steps = Vec::with_capacity(n);
+                let mut wall_times = Vec::with_capacity(n);
+                let mut values = Vec::with_capacity(n);
+                for (step, wall_time, &BlobSequenceValue(ref value)) in points {
+                    steps.push(step.into());
+                    wall_times.push(wall_time.into());
+                    let eid = req.experiment_id.as_str();
+                    let blob_refs = (0..value.len())
+                        .map(|i| {
+                            let bk = BlobKey {
+                                experiment_id: Cow::Borrowed(eid),
```

Comment:
Interesting; I should have asked this question before but when does it turn into mutable (`to_mut`)? It is not 100% obvious to me reading the blob_key.rs.

Reply:
Asking now is fine! :-) Depending on exactly what you’re asking, I think
maybe the answer is “when we call `bk.to_string()` below”? but let me
provide some more context.

Blob keys have string fields defined as `Cow<'a, str>` because we can
get blob keys in two different ways. First, in `ReadBlobSequences`, we
synthesize blob keys from the runs and tags that we already have in
memory. Second, in `ReadBlob`, we parse a blob key from user input.
When we parse a blob key from a base64 string, we have to allocate new
memory for the eid/run/tag: no way around it, since they’re not stored
anywhere else. But when we’re synthesizing a blob key, we’d like to
avoid cloning the eid/run/tag, since we already have references to them.
This is where the [`Cow<'a, str>`] comes in: you can think of it as
“either an owned string or a borrowed string, plus a bit indicating
which is which so that we know whether to free memory when this is
dropped”.

So on this line, we create a blob key that just points into the existing
string structures. Then, when we create the blob reference below, we
encode it all as base64-over-JSON, which allocates a new `String`.

Maybe this is what you meant by “turn into mutable (`to_mut`)”, or maybe
not; does this help? if not, can you clarify?

[`Cow<'a, str>`]: https://doc.rust-lang.org/std/borrow/enum.Cow.html

Reply:
Thanks for the explanation. I think you are answering most of my question. However, my confusions really lies with the basic of string manipulation. For instance, when invoking `bk.to_string()`, I suspect it would be "concatenating" `run`, `tag`, `index`, and `step` and when doing so, why you would have to "clone". I definitely see why you would have to allocate a new space to store the new serialized string but I am not sure why that's perceived as cloning, for example, `run` and others.

Reply:
> when invoking `bk.to_string()`, I suspect it would be "concatenating" `run`, `tag`, `index`, and `step` and when doing so, why you would have to "clone". I definitely see why you would have to allocate a new space to store the new serialized string but I am not sure why that's perceived as cloning, for example, `run` and others.

Right. You don’t have to *clone* `run`, but you do have to memcpy its
contents into a new serialized string.

By “when we’re synthesizing a blob key, we’d like to avoid cloning the
eid/run/tag”, I mean that we don’t want to do this:

```rust
let bk = BlobKey {
    experiment_id: eid.clone(),
    run: run.clone(),
    // ...
};
```

—which would be a literal `clone`—and so instead we use `Cow<'a, str>`s
so that the blob key can just borrow the data.